Makefile

```diff
@@ -20,14 +27,19 @@ check-formatting: generate deps vet lint
 
 # Build manager binary
 manager: generate fmt vet
-	go build -o bin/manager github.com/kudobuilder/kudo/cmd/manager
+	go build -ldflags "-X ${GIT_VERSION_PATH}=${GIT_VERSION} \
+             	-X ${GIT_COMMIT_PATH}=${GIT_COMMIT} \
+             	-X ${BUILD_DATE_PATH}=${BUILD_DATE}" \
+             	-o bin/manager github.com/kudobuilder/kudo/cmd/manager
 
 # Run against the configured Kubernetes cluster in ~/.kube/config
 run: generate fmt vet
-	go run ./cmd/manager/main.go
+	go run -ldflags "-X ${GIT_VERSION_PATH}=${GIT_VERSION} \
```

Comment:
Will we be hosting an endpoint at to be able to have the CLI query the server version like the `kubectl version` call does?  It looks like we provide this so we can check the logs only.

Reply:
agreeing with you for the UX.

```
client -> kudo cli version
server -> kudo controller version
```

For now I see a transitional state so we have at least something via `pkg/version`. 

Ideally imo we should consider having it organized in `pkg/kudoctl/version` for the CLI and `pkg/controller/version` for the controller version. I think this also relates as one part of the puzzle here https://github.com/kudobuilder/kudo/issues/163 and should be considered when deciding how to move forward.

Reply:
The `kubectl version` and `helm version` behavior is to report both (if the server is down, the client still runs and the CLIs let the server time out). Is this something we should match?

Reply:
I guess one reason of why it was implemented that way is, that `kubectl` or `helm` cli versions don't know at compile time on which cluster they are running. They couldn't provide that information without talking to outside their constraints. The cli itself does not provide you insides of an external component, e.g. in our case the KUDO controller. The only way of getting this controller version is calling an endpoint because that tells the controller version otherwise it would just be wild guessing. This also helps when you have e.g. multiple namespaces with different KUDO controller versions installed. Of course the controller itself should display its version in the log messages at startup, but providing it to any tool outside of it which is interested seems like a good idea to match for me. 

Now the question for me is if we should just open any reasonable port that listens for a `get` event through a `/version` route OR is there another way how we could provide to the outside world from a controller perspective this information. I could think for instance of incorporating this as a Prometheus exposed metric among other metrics that could lead a KEP for monitoring and logging. Finding the right [port](https://github.com/prometheus/prometheus/wiki/Default-port-allocations) that exposes in Prometheus format metrics such as:
- Version number
- Reconciliation counts
- Deployment counts
- Errortype counts
- etc.

That would though mean that this endpoint is available from within the cli and could also add more attack surface. Also think about the scenario when you run your controller outside the cluster. How would the cli know to reach the endpoint? Another way of dealing with this would be some sort of KUDO controller registry (cluster wide or namespaced) that is a CRD or configmap that holds all important metadata of a KUDO controller and everytime a controller starts it would update this information. During the startup/boottime of the KUDO controller it would then update this registry and the `kubectl kudo version` command would just need to access this registry to get the data, e.g. accessing a data point in a configmap or crd.


One thing I don't like with `kubectl version` and reporting both (client but also on the server side calling an api) is the timeout and wait time. If e.g. your pointed to a cluster that is down. Imo thats a bad user experience. If I stop my minikube it would take a while and show me just combined the outpout:

```bash
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.2", GitCommit:"cff46ab41ff0bb44d8584413b598ad8360ec1def", GitTreeState:"clean", BuildDate:"2019-01-13T23:15:13Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"darwin/amd64"}
Unable to connect to the server: dial tcp 192.168.64.12:8443: i/o timeout
```

That happened after an entire (sure timeouts can be changed but it felt very long). The client version should show immediately its version (there is no need to call an api for that as we can add this information during compile time), and the timeout message can show after the timeout. With kubectl as it is right now the timeout blocks both.